src/C/sahap-c.c

```diff
@@ -148,6 +152,188 @@ static void InitializeSystem(void) {
     }
 }
 
+
+// NEW CODE STARTS HERE: 
+
+
+// Extract k-mer from the tail of a read (null-terminated)
+char* tailKmer(const READ* r, int k) {
+    if (!r || !r->let || k > r->numSites) return NULL;
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    int startIdx = r->numSites - k;
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[startIdx + i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Extract k-mer from the head of a read (null-terminated)
+char* headKmer(const READ* r, int k) {
+    if (!r || !r->let || k > r->numSites) return NULL;
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Check if two k-mers match (string compare)
+bool kmersMatch(const char* kmer1, const char* kmer2) {
+    if (!kmer1 || !kmer2) return false;
+    return strcmp(kmer1, kmer2) == 0;
+}
+
+// Compare two reads for equality
+bool readsEqual(const READ* a, const READ* b) {
+    if (a->numSites != b->numSites) return false;
+    if (a->firstSite != b->firstSite) return false;
+    for (int i = 0; i < a->numSites; ++i) {
+        if (a->let[i] != b->let[i]) return false;
+    }
+    return true;
+}
+
+// Append one base to the right of a read
+READ addReadsRight(const READ* base, char a) {
+    READ out = *base; 
+    out.numSites = base->numSites + 1;
+
+    char* newLet = (char*)malloc(out.numSites * sizeof(char));
+    if (!newLet) {
```

Comment:
The only way newLet can be NULL is if malloc returns NULL... which is a fatal error (program should die, not return an invalid out variable).